File: java/src/main/java/challenges/aoc2024/day3/model/ComputerMemorySection.java
package challenges.aoc2024.day3.model;

import java.util.List;
import java.util.regex.Matcher;

import org.immutables.value.Value.Immutable;

import challenges.aoc2024.day3.util.DayThreeConstants;

@Immutable
public interface ComputerMemorySection {
    
    List<MultiplicationInstruction> getInstructions();

    static ComputerMemorySection of(String rawMemory) {
        final Matcher matcher = DayThreeConstants.MULTIPLICATION_INSTRUCTION_PATTERN.matcher(rawMemory);
        
        return ImmutableComputerMemorySection.builder()
            .instructions(matcher.results()
                .map(match -> match.group(0))
                .map(MultiplicationInstruction::of)
                .toList())
            .build();
    }
}

File: scala/src/main/scala/challenges/aoc2024/day3/model/ComputerMemorySection.scala
package challenges.aoc2024.day3.model

import challenges.aoc2024.day3.util.DayThreeConstants

import scala.util.matching.Regex

case class ComputerMemorySection(instructions: List[MultiplicationInstruction])

object ComputerMemorySection {

  def of(rawMemory: String): ComputerMemorySection = {

    val regex: Regex = DayThreeConstants.multiplicationInstructionRegex

    ComputerMemorySection(
      regex.findAllMatchIn(rawMemory)
        .map(m => MultiplicationInstruction.of(m.group(0)))
        .toList
    )
  }
}
